### PR TITLE
add 2d free fermion mpo

### PR DESCRIPTION
**Pull request by mganahl**

### CODE REVIEW

File: tensornetwork/matrixproductstates/mpo.py [hunk 1]
```diff
@@ -286,3 +286,102 @@ def __init__(self,
     temp[2, 0, :, :] = self.Bz[-1] * sigma_z
     mpo.append(temp)
     super().__init__(tensors=mpo, backend=backend, name=name)
+
+
+class FiniteFreeFermion2D(FiniteMPO):
+  """
+  Free fermsions on a 2d grid
```
Comment (jacksonwb) [inline tensornetwork/matrixproductstates/mpo.py:293]: typo

File: tensornetwork/matrixproductstates/mpo_test.py [hunk 2]
```diff
@@ -5,7 +5,26 @@
...
+  neighbors = {}
+  mat = np.arange(N1 * N2).reshape(N1, N2)
+  for n in range(N1 * N2):
+    x, y = np.divmod(n, N2)
+    if n not in neighbors:
+      neighbors[n] = []
+    if y < N2 - 1:
+      neighbors[n].append(mat[x, y + 1])
+    if x > 0:
+      neighbors[n].append(mat[x - 1, y])
```
Comment (jacksonwb) [inline tensornetwork/matrixproductstates/mpo_test.py:26]: This only needs to find forward adjacency?

File: tensornetwork/matrixproductstates/mpo.py (hunk 1 above)
Comment (jacksonwb) [inline tensornetwork/matrixproductstates/mpo.py:293]: typo

File: tensornetwork/matrixproductstates/mpo_test.py (hunk 2 above)
Comment (jacksonwb) [inline tensornetwork/matrixproductstates/mpo_test.py:26]: Unless this is useful elsewhere perhaps place it inside `test_finiteFreeFermions2d`

File: tensornetwork/matrixproductstates/mpo.py [hunk 3]
```diff
@@ -286,3 +286,102 @@ def __init__(self,
...
+
+      mpo_matrix[2 * N1 + 1, 0, :, :] = v * particle_number
+
+      mpo_matrix[2 * N1 + 1, 1, :, :] = _t1 * cdag # cdag @ sigma_z == cdag
+      mpo_matrix[2 * N1 + 1, N1, :, :] = _t2 * cdag
+      mpo_matrix[2 * N1 + 1, N1+1, :, :] = _t1 * c # c @ sigma_z == -c
+      mpo_matrix[2 * N1 + 1, 2*N1, :, :] = _t2 * c # c @ sigma_z == -c
+
+      mpo_matrix[2 * N1 + 1, 2 * N1 + 1, :, :] = eye
+      mpo.append(mpo_matrix.copy())
```
Comment (jacksonwb) [inline tensornetwork/matrixproductstates/mpo.py:379]: Is the `.copy()` needed? Seems like the reasign with `np.zeros` at the start of each loop should be sufficient no?

**Review by jacksonwb — APPROVED** [review]

File: tensornetwork/matrixproductstates/mpo_test.py (hunk 2 above)
Comment (mganahl) [reply tensornetwork/matrixproductstates/mpo_test.py:26]: that's right

File: tensornetwork/matrixproductstates/mpo.py (hunk 3 above)
Comment (mganahl) [reply tensornetwork/matrixproductstates/mpo.py:379]: removed

File: tensornetwork/matrixproductstates/mpo_test.py (hunk 2 above)
Comment (mganahl) [reply tensornetwork/matrixproductstates/mpo_test.py:26]: done

**Pull request merged**